### PR TITLE
[201911][Monit] Monitor critical processes in radv and dhcp_relay containers.

### PR DESCRIPTION
**Pull request by yozhao101**
Signed-off-by: Yong Zhao <yozhao@microsoft.com>

#### Why I did it
This PR aims to monitor critical processes in `router advertiser` and `dhcp_relay` containers by Monit.

#### How I did it
Router advertiser container only ran on T0 device and the T0 device should have at least one VLAN interface
which was configured an IPv6 address. At the same time, router advertiser container will not run on devices of which
the deployment type is 8.

As such, I created a service which will dynamically generate Monit configuration file of router advertiser from a
template.

Similarly Monit configuration file of `dhcp_relay` was also generated from a template since the number of `dhcrelay` process in `dhcp_relay` container is depended on number of VLANs.

#### How to verify it
I verified this implementation on a DuT.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [x ] 201911
- [ ] 202006
- [ ] 202012

### CODE REVIEW

File: rules/docker-dhcp-relay.mk [hunk 1]
```diff
@@ -26,3 +26,4 @@ $(DOCKER_DHCP_RELAY)_CONTAINER_NAME = dhcp_relay
 $(DOCKER_DHCP_RELAY)_RUN_OPT += --privileged -t
 $(DOCKER_DHCP_RELAY)_RUN_OPT += -v /etc/sonic:/etc/sonic:ro
 $(DOCKER_DHCP_RELAY)_FILES += $(SUPERVISOR_PROC_EXIT_LISTENER_SCRIPT)
+$(DOCKER_DHCP_RELAY)_BASE_IMAGE_FILES += monit_dhcp_relay.j2:/usr/share/sonic/tempaltes
```
Comment (jleveque) [inline rules/docker-dhcp-relay.mk:29]: Typo: s/tempaltes/templates/

File: rules/docker-router-advertiser.mk [hunk 2]
```diff
@@ -26,3 +26,4 @@ $(DOCKER_ROUTER_ADVERTISER)_CONTAINER_NAME = radv
 $(DOCKER_ROUTER_ADVERTISER)_RUN_OPT += --privileged -t
 $(DOCKER_ROUTER_ADVERTISER)_RUN_OPT += -v /etc/sonic:/etc/sonic:ro
 $(DOCKER_ROUTER_ADVERTISER)_FILES += $(SUPERVISOR_PROC_EXIT_LISTENER_SCRIPT)
+$(DOCKER_ROUTER_ADVERTISER)_BASE_IMAGE_FILES += monit_radv.j2:/usr/share/sonic/tempaltes
```
Comment (jleveque) [inline rules/docker-router-advertiser.mk:29]: Typo: s/tempaltes/templates/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: rules/docker-dhcp-relay.mk (hunk 1 above)
Comment (yozhao101) [reply rules/docker-dhcp-relay.mk]: Good catch! Fixed.

**Review by jleveque — APPROVED** [review]

File: dockers/docker-dhcp-relay/base_image_files/monit_dhcp_relay.j2 [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      {%- if VLAN and vlan_name in VLAN and 'dhcp_servers' in VLAN[vlan_name] and VLAN[vlan_name]['dhcp_servers']|length >0 -%}
+        {%- for dhcp_server in VLAN[vlan_name]['dhcp_servers'] -%}
+          {%- if dhcp_server | ipv4 -%}
+            {%- set relay_for_ipv4.flag = True -%}
+          {%- endif -%}
+        {%- endfor -%}
+        {%- if relay_for_ipv4.flag -%}
+          {%- set relay_for_ipv4 = False -%}
+          {# Check the running status of each DHCP relay agent instance #}
+check program dhcp_relay|dhcrelay with path "/usr/bin/process_checker dhcp_relay /usr/sbin/dhcrelay -d -m discard -a %h:%p %P --name-alias-map-file /tmp/port-name-alias-map.txt -id {{ vlan_name }}"
```
Comment (yozhao101) [inline dockers/docker-dhcp-relay/base_image_files/monit_dhcp_relay.j2:30]: @jleveque At here, I used the argument `-a %h:%p P%`. However, the template file of superviord of `dhcp_relay` container which used ` -a %%h:%%p %%P`: https://github.com/Azure/sonic-buildimage/blob/ee728aab7b921ea9d0e81dda1de3a6f0279397fd/dockers/docker-dhcp-relay/docker-dhcp-relay.supervisord.conf.j2#L78. In `dhcp_relay` container, the command line of `dhcrelay` process used `-a %h:%p P%`. If I used ` -a %%h:%%p %%P` in this template file, `sonic-cfggen` will generate ` -a %%h:%%p %%P` not ` -a %h:%p %P`.

Comment (jleveque) [reply dockers/docker-dhcp-relay/base_image_files/monit_dhcp_relay.j2:30]: Yes, that's correct. In the supervisor config file, we need to escape the percent sign with an extra percent sign.

**Comment by rlhui** [comment]
could we please paste what's the corresponding sonic-mgmt test case for this?

**Comment by yozhao101** [comment]
> could we please paste what's the corresponding sonic-mgmt test case for this?

I am working on the PR of pytest script and will paste the link of PR later.

**Pull request merged**